### PR TITLE
add vsyscall table in elf runtime memory

### PR DESCRIPTION
**Pull request by ucgJhe**
as title. so we can run hello world golang binary now.
if golang binary exmaple should be included in this PR please let me know.

### CODE REVIEW

File: qiling/loader/elf.py [hunk 1]
```diff
@@ -615,4 +615,23 @@ def load_with_ld(self, stack_addr, load_address = -1, argv = [], env = {}):
         self.new_stack = new_stack
         self.load_address = load_address
         self.images.append(self.coverage_image(load_address, load_address + mem_end, self.path))
-        self.ql.os.function_hook = FunctionHook(self.ql, self.elf_phdr + mem_start, self.elf_phnum, self.elf_phent, load_address, load_address + mem_end)
\ No newline at end of file
+        self.ql.os.function_hook = FunctionHook(self.ql, self.elf_phdr + mem_start, self.elf_phnum, self.elf_phent, load_address, load_address + mem_end)
+
+        # map vsyscall section for some specific needs
+        _vsyscall_addr = 0xffffffffff600000
```
Comment (xwings) [inline qiling/loader/elf.py:621]: this address is very "x8664 ?"

File: qiling/loader/elf.py [hunk 2]
```diff
@@ -615,4 +615,23 @@ def load_with_ld(self, stack_addr, load_address = -1, argv = [], env = {}):
...
+            self.ql.mem.map(_vsyscall_addr, _vsyscall_size, info="[vsyscall]")
+            self.ql.mem.write(_vsyscall_addr, _vsyscall_size * b'\xcc')
+
+            _vsyscall_entry_asm = [0x60c0c748, # gettimeofday
+                                   0xc9c0c748, # time
+                                   ]
+
+            for idx, val in enumerate(_vsyscall_entry_asm):
+                self.ql.mem.write(_vsyscall_addr + idx * 0x400 + (idx+0) * 4, self.ql.pack32(val))        # entry
+                self.ql.mem.write(_vsyscall_addr + idx * 0x400 + (idx+1) * 4, self.ql.pack64(0x0f000000)) # syscall
```
Comment (xwings) [inline qiling/loader/elf.py:636]: any reason to fix all these address ?

Comment (ucgJhe) [reply qiling/loader/elf.py:636]: vsyscall table should be looks like this in memory
image

**Comment by aquynh** [comment]
The question is why hardcode them in the loader?

File: qiling/loader/elf.py (hunk 1 above)
Comment (ucgJhe) [reply qiling/loader/elf.py:621]: you are right , its only for x8664 elf binary

**Comment by ucgJhe** [comment]
> The question is why hardcode them in the loader?

so we can run golang binary without extra setup

any advice is welcome

**Comment by aquynh** [comment]
I mean these addresses never change?

**Comment by ucgJhe** [comment]
> I mean these addresses never change?

I check  it here ref: http://lxr.linux.no/linux-bk+v2.5.5/include/asm-x86_64/vsyscall.h#L10
depends on how linux kernel do it , i will check other version make sure its fixed

**Comment by aquynh** [comment]
At least use profile for the hardcode values, as always.

**Comment by ucgJhe** [comment]
> At least use profile for the hardcode values, as always.

got it, ill move it to profile 

thanks

File: qiling/loader/elf.py [hunk 3]
```diff
@@ -615,4 +615,24 @@ def load_with_ld(self, stack_addr, load_address = -1, argv = [], env = {}):
...
+            _vsyscall_addr = int(self.ql.os.profile.get("OS64", "vsyscall_address"), 16)
+            _vsyscall_size = int(self.ql.os.profile.get("OS64", "vsyscall_size"), 16)
+
+            if not self.ql.mem.is_mapped(_vsyscall_addr, _vsyscall_size):
+                # initialize with \xcc then insert syscall entry
+                # each syscall should be 1KiB(0x400 bytes) away
+                self.ql.mem.map(_vsyscall_addr, _vsyscall_size, info="[vsyscall]")
+                self.ql.mem.write(_vsyscall_addr, _vsyscall_size * b'\xcc')
+
+                _vsyscall_entry_asm = [0x60c0c748, # gettimeofday
```
Comment (xwings) [inline qiling/loader/elf.py:631]: i assume this is opcode and can we use ql.compile here ?

Comment (ucgJhe) [reply qiling/loader/elf.py:631]: its doable , but I found that the result of ql.compile need to be adjusted to fit the size, will make it more complicated

Comment (xwings) [reply qiling/loader/elf.py:631]: try if possible. so code will be more readable. else stick to this is fine.

Comment (ucgJhe) [reply qiling/loader/elf.py:631]: i did try , doesnt look good to me. I recommend stick to this , and i will comment the asm after each one, is that acceptable ?

Comment (ucgJhe) [reply qiling/loader/elf.py]: just figured out how to use ql.compile correctly, it should be more readable now.

**Comment by xwings** [comment]
ok, now we support golang binary. cool!

**Pull request merged**